### PR TITLE
fix iterator in freeze_lazy_contraints

### PR DESCRIPTION
**Pull request by polgreen**
see https://github.com/diffblue/cbmc/issues/1581

### CODE REVIEW

File: src/solvers/refinement/refine_arrays.cpp [hunk 1]
```diff
@@ -130,7 +130,7 @@ void bv_refinementt::freeze_lazy_constraints()
     for(std::set<symbol_exprt>::const_iterator it=symbols.begin();
```
Comment (romainbrenguier) [inline src/solvers/refinement/refine_arrays.cpp]: While we are changing this loop, can we make it a ranged-for? (and the previous loop as well)

File: src/solvers/refinement/refine_arrays.cpp [hunk 2]
```diff
@@ -130,7 +130,7 @@ void bv_refinementt::freeze_lazy_constraints()
...
         it!=symbols.end(); ++it)
     {
-      bvt bv=convert_bv(l_it->lazy);
+      bvt bv=convert_bv(*it);
```
Comment (romainbrenguier) [inline src/solvers/refinement/refine_arrays.cpp]: should be const

**Review by romainbrenguier — APPROVED** [review]
The code of the function could be improved, but this should not block the fix.

**Review by martin-cs — APPROVED** [review]
Yes if it has been tested with --refine.

**Comment by polgreen** [comment]
Are there regression tests for --refine?  It's been testing with --refine-strings, because there seem to be plenty of regression tests for that. 

I'm not really sure what the expected behaviour of --refine should be, so I'm not in a position to write regression tests for it.

**Comment by martin-cs** [comment]
On Mon, 2017-11-20 at 16:52 +0000, E Polgreen wrote:
> Are there regression tests for --refine?

It appears not :-(

>   It's been testing with --refine-strings, because there seem to be plenty of regression tests for that. 

That's a good start.

> I'm not really sure what the expected behaviour of --refine should be, so I'm not in a position to write regression tests for it. 

It should only affect performance, not correctness.  It will slightly
change the status messages but it shouldn't affect the actual results.
It /should/ mean that the solver runs faster when the code contains
multiply and divide operations.

**Comment by polgreen** [comment]
I've run this on the tests in regression/cbmc with --refine used as an option. 

There is 1 test that fails with a seg fault, which is Float-div1. This test also fails if I run it using develop and --refine

Perhaps someone who is familiar with the refinement stuff could debug that, as it's not a bug introduced by this PR. 

Would it also be reasonable to add a run of the tests in the cbmc folder using the "--refine" option to the CI?

**Comment by tautschnig** [comment]
#1621, which now addresses #1620, also adds a first regression test using `--refine`.

**Comment by tautschnig** [comment]
@polgreen Would you mind following @romainbrenguier's suggestions for some cleanup?

**Comment by tautschnig** [comment]
Also, would you mind adding "Fixes: #1581" to the commit message?

File: src/solvers/refinement/refine_arrays.cpp [hunk 3]
```diff
@@ -121,16 +121,13 @@ void bv_refinementt::freeze_lazy_constraints()
   if(!lazy_arrays)
     return;
 
-  for(std::list<lazy_constraintt>::iterator
-        l_it=lazy_array_constraints.begin();
-      l_it!=lazy_array_constraints.end(); ++l_it)
+  for(const auto &l_it : lazy_array_constraints)
```
Comment (tautschnig) [inline src/solvers/refinement/refine_arrays.cpp]: Nit picking: may I suggest renaming as it isn't an iterator anymore?

**Review by smowton — COMMENTED** [review]
Add a regression test exposing the bug?

**Comment by polgreen** [comment]
I think that is out of scope for this PR, and should be done by a diffblue employee who is familiar with the refine functionality. Refine should probably have more than one regression test anyway.

I only created this PR because it seemed like it would be straightforward and quick to do given that @kroening has already given the fix on this issue report. I didn't come across any case where the bug actually affected analysis, I just spotted an unused iterator in the code: https://github.com/diffblue/cbmc/issues/1581

**Comment by peterschrammel** [comment]
@polgreen, there are a bunch of tests in https://github.com/diffblue/cbmc/tree/develop/regression/array-refinement that use that functionality but are not run at the moment.

**Comment by polgreen** [comment]
@peterschrammel those use "refine-arrays" but not "refine". I'm not sure what the difference is, but @martin-cs mentioned this should be run with "--refine", and that was the option I was referring to when I said refine should have more than one regression test. Do "refine" and "refine-arrays" do the same thing?

**Comment by peterschrammel** [comment]
Grep for 'refine' in https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_parse_options.cpp and
https://github.com/diffblue/cbmc/blob/develop/src/cbmc/cbmc_solvers.cpp to see how it works. 
--refine subsumes --refine-arrays, the code that you fixed is exercised by --refine-arrays.

**Pull request merged**